### PR TITLE
Add RPC internal helper that overrides the default pickler.

### PR DESCRIPTION
**Pull request by zzzwen**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30185 Add RPC internal helper that overrides the default pickler.**

To enable share_memory over RPC, add an internal helper that overrides the default RPC pickler.
Replace D18598974

Differential Revision: [D18621372](https://our.internmc.facebook.com/intern/diff/D18621372/)

### CODE REVIEW

**Comment by zzzwen** [comment]
Replaced https://github.com/pytorch/pytorch/pull/30173

File: test/rpc_test.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 from common_utils import load_tests
 import dist_utils
 from dist_utils import dist_init
+from torch.distributed.rpc.api import _use_rpc_pickler
```
Comment (pritamdamania87) [inline test/rpc_test.py]: nit: can we put this in `torch.distributed.rpc` and not `torch.distributed.rpc.api`?

**Review by pritamdamania87 — APPROVED** [review]
@pietern @satgera We're going with this approach in the very short term to unblock @zzzwen since there are some issues he is running into with `_apply` and we need to complete the integration soon. We'll try to resolve the issues with the `_apply` method in the mean time and remove this API eventually.

**Review by mrshenli — APPROVED** [review]
Let's land this. I will rebase #30217 when this is in.

**Comment by pietern** [comment]
@pritamdamania87 Thanks for clarifying, makes sense. Can you file an issue so we don't lose track of removing this eventually? Retaining this is a liability in my opinion.

**Comment by pritamdamania87** [comment]
@pietern Filed https://github.com/pytorch/pytorch/issues/30633

**Pull request closed without merging**